Displaydollsynchook

Patrikkk opened the pull request:
**The event:**
An event that is called when a player modifies the slot of a DisplayDoll (Mannequin).

I was trying to think from a developer-friendly perspective here.
Instead of passing separate variables for type/stack/prefix, thought I pass an Item object.
As well as, instead of having devs who work with this hook figure out and implement how to get the Item of the DisplayDoll, I just provide it in the hook.
I can imagine this being used for creative purposes in plugins.

**The handler:**
In a previous PR I have added the tile entity request packet handler which checks for building permissions to prevent the unauthorized player to open a DisplayDoll and see its content.

This Handler is being added to prevent *Hackers* from modifying a DisplayDoll through direct/crafted packet sending, or by sending raw byte data to the server.
In a valid environment, the player couldn't even get to see the content of the doll in the first place, to then try to modify its items.

Because of this,  I do not bother with making sure the player gets their item back.

Patrikkk committed:
Adding DisplayDollItemSync event.

An event that is called when a player modifies the slot of a DisplayDoll (Mannequin).

I was trying to think from a developer friendly perspective here.
Instead of passing seperate variables for type/stack/prefix, thought I pass an Item object.
As well as, instead of having devs who work with this hook figure out and implement how to get the Item of the DisplayDoll, I just provide it in the hook.
I can imagine this being used for creative purposes in plugins.

Patrikkk committed:
Add TEDisplayDoll object to the event args.

Patrikkk committed:
Add DisplayDollItemSyncHandler

In a previous PR I have added the tile entity request packet handler which checks for building permissions to prevent the unauthorized player to open a DisplayDoll and see its content.

This Handler is being added to prevent *Hackers* from modifying a DisplayDoll through direct/crafted packet sending, or by sending raw byte data to the server.
In a valid enviroment, the player couldn't even get to see the content of the doll in the first place, to then try to modify it's items.

Because of this,  I do not bother with making sure the player gets their item back.

Patrikkk committed:
Update CHANGELOG.md

Loggin latest change of adding DisplayDollItemSyncHandler.
Updating missing author for previous changes.

Patrikkk committed:
Moving the DisplayDollItemSync hook inside null check.

In crafted (hacked) packet sending, people could send the DisplayDollItemSync packet with an entity ID that is not actually a display doll. This would not happen in a normal scenario.
If they send the crafted packet, our hook would have been invoked and it would have contain a null TEDisplayDoll object since we tried to get the TileEntity object as a TEDisplayDoll object.

QuiCM commented on TShockAPI/GetDataHandlers.cs:
```diff
@@ -3703,6 +3762,48 @@ private static bool HandleEmoji(GetDataHandlerArgs args)
+				newItem.stack = stack;
+				newItem.Prefix(prefix);
+
+				if (oldItem.type == 0 && newItem.type == 0)
```
Can this check occur before calling SetDefaults?

QuiCM replied:
Maybe also check for `stack = 0` somewhere

QuiCM replied:
Can `oldItem` be null?

Patrikkk replied:
If stack is 0, nothing happens, no item is placed.
`oldItem` is never null because the doll entity has default Item() object set for empty slots.
Sent commit for the type check order.

Patrikkk committed:
Register DisplayDollItemSyncHandler.

Moved the type check for 0 before itemSetDefaults to have less code run if the empty slots are clicked.

Patrikkk committed:
Merge branch 'general-devel' into displaydollsynchook
